fearless/Modules/Staking/StakingAmount/ViewModel/BalanceViewModelFactory.swift

```diff
@@ -0,0 +1,135 @@
+import Foundation
+import SoraFoundation
+import IrohaCrypto
+import CommonWallet
+
+protocol BalanceViewModelFactoryProtocol {
+    func priceFromAmount(_ amount: Decimal, priceData: PriceData) -> LocalizableResource<String>
+    func amountFromValue(_ value: Decimal) -> LocalizableResource<String>
+    func balanceFromPrice(_ amount: Decimal, priceData: PriceData?)
+    -> LocalizableResource<BalanceViewModelProtocol>
+    func createBalanceInputViewModel(_ amount: Decimal?) -> LocalizableResource<AmountInputViewModelProtocol>
+    func createAssetBalanceViewModel(_ amount: Decimal, balance: Decimal?, priceData: PriceData?)
+    -> LocalizableResource<AssetBalanceViewModelProtocol>
+}
+
+final class BalanceViewModelFactory: BalanceViewModelFactoryProtocol {
+    let walletPrimitiveFactory: WalletPrimitiveFactoryProtocol
+    let selectedAddressType: SNAddressType
+
+    private lazy var formatterFactory = AmountFormatterFactory()
+    private lazy var priceAsset = {
+        walletPrimitiveFactory.createPriceAsset()
+    }()
+
+    private lazy var targetAsset = {
+        walletPrimitiveFactory.createAssetForAddressType(selectedAddressType)
+    }()
+
+    init(walletPrimitiveFactory: WalletPrimitiveFactoryProtocol, selectedAddressType: SNAddressType) {
+        self.walletPrimitiveFactory = walletPrimitiveFactory
+        self.selectedAddressType = selectedAddressType
+    }
+
+    func priceFromAmount(_ amount: Decimal, priceData: PriceData) -> LocalizableResource<String> {
+        guard let rate = Decimal(string: priceData.price) else {
+            return LocalizableResource { _ in "" }
+        }
+
+        let targetAmount = rate * amount
+
+        let priceAsset = walletPrimitiveFactory.createPriceAsset()
+        let localizableFormatter = formatterFactory.createTokenFormatter(for: priceAsset)
+
+        return LocalizableResource { locale in
+            let formatter = localizableFormatter.value(for: locale)
+            return formatter.string(from: targetAmount) ?? ""
+        }
+    }
+
+    func amountFromValue(_ value: Decimal) -> LocalizableResource<String> {
+        let localizableFormatter = formatterFactory.createTokenFormatter(for: targetAsset)
+
+        return LocalizableResource { locale in
+            let formatter = localizableFormatter.value(for: locale)
+            return formatter.string(from: value) ?? ""
+        }
+    }
+
+    func balanceFromPrice(_ amount: Decimal, priceData: PriceData?)
+    -> LocalizableResource<BalanceViewModelProtocol> {
+        let localizableAmountFormatter = formatterFactory.createTokenFormatter(for: targetAsset)
```

Comment:
Looks like these three methods could be reused in other parts. Should we make a helper instead?

Reply:
Yep, the idea is just to reuse the factory itself